### ISSUE TITLE
Redirect error in Typo 10.4.17

### ISSUE THREAD

**Issue by trokohl** [FluidTYPO3/flux#1893]
Hi,

i'm not sure if this is a typo3 or flux problem, but if i update to typo 10.4.17 and i have a redirect like

#^/abc(.*)$#

and call an url like https://example.com/abc123

i get the following error:

TypoScriptFrontendController was tried to be injected before initial creation

with this stacktrace:
/typo3/sysext/frontend/Classes/Controller/TypoScriptFrontendController.php line 4032
/typo3temp/var/cache/code/di/DependencyInjectionContainer_4b37854b95ecab2ed901a2359bd7a258866e3d88.php line 9756
/vendor/symfony/dependency-injection/Container.php line 248
/vendor/symfony/dependency-injection/Container.php line 228
/typo3/sysext/core/Classes/Utility/GeneralUtility.php line 3470
/typo3/sysext/extbase/Classes/Mvc/Web/Routing/UriBuilder.php line 178
/typo3/sysext/extbase/Classes/Object/Container/Container.php line 252
/typo3/sysext/extbase/Classes/Object/Container/Container.php line 169
/typo3/sysext/extbase/Classes/Object/Container/Container.php line 111
/typo3/sysext/extbase/Classes/Object/ObjectManager.php line 114
/typo3conf/ext/flux/Classes/Provider/AbstractProvider.php line 836
/typo3conf/ext/flux/Classes/Provider/AbstractProvider.php line 382
/typo3conf/ext/flux/Classes/Provider/AbstractProvider.php line 283
...

typo3 or flux?

thanks

**Comment by trokohl** [comment]
I can narrow it down a little more: 

The error only occurs when there is a Flux element on the target page.
Also, the error is not always reproducible, which was mostly due to the browser (redirect) and/or Typo3 cache.

**Comment by datamints-off** [comment]
The change in TYPO3 core that causes the error is this one: https://github.com/TYPO3/TYPO3.CMS/commit/14e0c88f01

**Comment by datamints-off** [comment]
I fixed the problem temporarily by adding the following code in \FluidTYPO3\Flux\Provider\AbstractProvider::getViewForRecord():

```
// Bad hack to fix https://github.com/FluidTYPO3/flux/issues/1893.
$callStack = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS);
$callStackFiltered = array_filter($callStack, function (array $item): bool {
    return array_key_exists('class', $item) && $item['class'] === \TYPO3\CMS\Redirects\Service\RedirectService::class;
 });
 if (count($callStackFiltered) > 0) {
    return GeneralUtility::makeInstance(ObjectManager::class)->get($viewClassName);
}
```

**Comment by derhansen** [comment]
I created an issue in TYPO3 core issue tracker (https://forge.typo3.org/issues/95118), since I believe this has to be fixed in TYPO3 core.

**Comment by derhansen** [comment]
What I asked myself: Is it required to set an `UriBuilder` instance to the rendering context at this place? In TYPO3 10.4.16, the created instance of the `UriBuilder` has no instance of `ContentObject`, so it is basically not possible to use the `UriBuilder` to create frontend links. I therefore tried to just remove the `UriBuilder` completely (since I could not find any other usage in the code of ext:flux) and obviously everything still works as expected.

Current Problem is, that this bug may block TYPO3 10.4 websites from applying the latest security updates, since redirects are broken due to the error.

Maybe @NamelessCoder can bring some light into this topic.

**Comment by NamelessCoder** [comment]
This would appear to be a bug introduced by making UriBuilder assume that an instance of TSFE can be injected - which isn't always the case and which definitely isn't the case in BE unless FE is simulated. Flux needs to be able to initialize UriBuilder in order to create a valid ControllerContext - which was safe to do before the core changed the UriBuilder constructor (but did have the side effect that you couldn't build a link with e.g. `f:link.action` inside the `Configuration` section of a Flux-enabled template; something that isn't a valid use case in the first place).

I'm unsure how to proceed with this and would like to know the standpoint of the core first, given that technically it is a breaking change for use cases of UriBuilder outside of late-stage or simulated FE implementations. Personally I wouldn't have changed this UriBuilder constructor but would rather have thrown an exception at places where UriBuilder attempts to access the ContentObjectRenderer but finds no instance.

**Comment by ohader** [comment]
What is the counter part in `ext:flux` relying on that behavior? Is that `AbstractProvider`?

**Comment by NamelessCoder** [comment]
It's AbstractProvider and if the call stack in OP's description is complete, it starts when getForm() is called - which happens for example when reading the DS (from, I assume, a pages record).

**Comment by derhansen** [comment]
Issue will be fixed with TYPO3 10.4.21 - see https://review.typo3.org/c/Packages/TYPO3.CMS/+/71120

**Comment by NamelessCoder** [comment]
Thanks for keeping track of this one, @derhansen - I'll close the issue and send a kind thought to Benjamin Franzke for fixing it in the TYPO3 core.
